### ISSUE TITLE
Manage threading in the backend

### ISSUE THREAD

**Issue by patrick-nicodemus** [PythonOT/POT#492]
This feature arguably falls outside the scope of the library - it is really a problem which is 2 dependencies down the food chain - but I will suggest it anyway because I think it makes a difference in quality of life for users. The linear algebra backend called by Numpy sometimes does very naive and inappropriate multithreading, where a single small matrix multiplication can be parallelized across all cores of the machine, resulting in a longer computation that takes more cores. I am running the computation on a 20 core machine, and if I try to compute the GW distances between two 100x100 matrices, it engages all 20 cores and takes 20ms; forcing Numpy to use a single thread brings the computation to about 12ms.

This problem is quite subtle to debug, as Numpy (or really its BLAS backend) launches 20 threads at the beginning of the program *no matter how large the matrices are*, but then it dynamically decides whether to use all threads or leave some idle. This is a clever design decision as it gets past the overhead of creating the threads up front, but for debugging parallelization-related performance issues it is quite confusing, because the system debugger will tell you that 20 threads are active, when in reality 1 is active and 19 are idle.

My feature request is that you integrate a tool such as joblib/threadpoolctl into the library and provide options to the user to control or limit thread-level parallelization, together with sensible defaults.

**Comment by Guddubhaiya07** [comment]
Best AI Model for Code. Beats ChatGPT

Toggle Sidebar
import numpy as np
from threadpoolctl import threadpool_limits

# Limit the number of threads used by NumPy to 1
with threadpool_limits(limits=1, user_api='blas'):
    a = np.random.rand(100, 100)
    b = np.random.rand(100, 100)
    c = np.dot(a, b)

￼

This feature arguably falls outside the scope of the library - it is really a problem which is 2 dependencies down the food chain - but I will suggest it anyway because I think it makes a difference in quality of life for users. The linear algebra backend called by Numpy sometimes does very naive and inappropriate multithreading, where a single small matrix multiplication can be parallelized across all cores of the machine, resulting in a longer computation that takes more cores. I am running the computation on a 20 core machine, and if I try to compute the GW distances between two 100x100 matrices, it engages all 20 cores and takes 20ms; forcing Numpy to use a single thread brings the computation to about 12ms.

This problem is quite subtle to debug, as Numpy (or really its BLAS backend) launches 20 threads at the beginning of the program no matter how large the matrices are, but then it dynamically decides whether to use all threads or leave some idle. This is a clever design decision as it gets past the overhead of creating the threads up front, but for debugging parallelization-related performance issues it is quite confusing, because the system debugger will tell you that 20 threads are active, when in reality 1 is active and 19 are idle.

My feature request is that you integrate a tool such as joblib/threadpoolctl into the library and provide options to the user to control or limit thread-level parallelization, together with sensible defaults.

￼

Your observation is correct that NumPy's linear algebra backend can sometimes use naive multithreading, which can lead to performance issues. This is particularly true for small matrix multiplications, where the overhead of creating threads can be larger than the computation itself.

The issue you're facing is a common one in parallel computing, where the overhead of creating threads can be larger than the computation itself. This is known as the "thread creation overhead" problem.

The issue you're facing is not unique to NumPy. It's a common problem in parallel computing, and it's not always straightforward to solve.

One approach to address this issue is to use a library like joblib/threadpoolctl that provides more control over the number of threads used. This can help you avoid the overhead of creating threads for small matrix multiplications.

However, it's important to note that this issue is not specific to NumPy. It's a general problem in parallel computing. As a result, it's not always possible to provide a one-size-fits-all solution.

That being said, we can consider adding an option to NumPy to control the number of threads used for parallel computations. This would allow users to control the trade-off between parallelism and thread creation overhead.

Here's a simple example of how you might use joblib/threadpoolctl to control the number of threads used by NumPy:

python

EditDownloadCopy code

1import numpy as np 2from threadpoolctl import threadpool_limits 3 4# Limit the number of threads used by NumPy to 1 5with threadpool_limits(limits=1, user_api='blas'): 6 a = np.random.rand(100, 100) 7 b = np.random.rand(100, 100) 8 c = np.dot(a, b)

In this example, we use the threadpool_limits context manager from the threadpoolctl library to limit the number of threads used by NumPy to 1. This can help reduce the overhead of creating threads for small matrix multiplications.

However, it's important to note that this approach may not always be suitable for every use case. In some cases, it may be more efficient to use multiple threads for parallel computations, even if it means creating more threads than necessary for small matrix multiplications.

As a result, it's important to carefully consider the trade-offs between parallelism and thread creation overhead when using libraries like NumPy.

In conclusion, while we can't guarantee a one-size-fits-all solution to this problem, we can consider adding an option to NumPy to control the number of threads used for parallel computations. This would allow users to control the trade-off between parallelism and thread creation overhead, and it would provide more flexibility for users who need to optimize their computations for specific use cases.
